source/rust_verify/src/singular.rs

```diff
@@ -339,67 +336,66 @@ pub fn check_singular_valid(
         panic!("internal error: integer_ring")
     };
 
-    let arc_ens: &air::ast::Stmt = stmts.last().unwrap();
-    let ens = match &**arc_ens {
-        air::ast::StmtX::Assert(error, exp) => (exp.clone(), error.clone()),
-        _ => {
-            panic!("internal error");
-        }
-    };
-
-    let mut vars: Vec<Ident> = vec![];
+    let mut vars: Vec<String> = vec![];
     for d in &**decl {
         if let air::ast::DeclX::Var(name, _typ) = &**d {
-            vars.push(name.clone());
+            let mut v2: String = name.to_string();
+            v2.push('@');
+            vars.push(sanitize_var_name(v2));
         }
     }
 
-    let mut reqs = vec![];
-    for idx in 0..(stmts.len() - 1) {
-        // last element is ensures clause
-        let stm = &stmts[idx];
-        match &**stm {
-            air::ast::StmtX::Assert(error, exp) => {
-                reqs.push((exp.clone(), error.clone()));
-            }
-            _ => {
-                panic!("internal error");
-            }
-        };
-    }
-
-    let query = match singular_printer(&vars, &reqs, &ens) {
-        Ok(query_string) => query_string,
-        Err(err) => return ValidityResult::Invalid(None, err),
-    };
-
-    air::singular_manager::log_singular(context, &query, func_span);
-
-    let quit_string = format!("{};", QUIT_SINGULAR);
-    let query = format!("{} {}", query, quit_string);
-
+    let mut encoder = SingularEncoder::new(vars);
     let singular_manager = SingularManager::new();
     let mut singular_process = singular_manager.launch();
-    let res = singular_process.send_commands(query.as_bytes().to_vec());
-    if (res.len() == 1) && (res[0] == "0") {
-        ValidityResult::Valid
-    } else if res[0].contains("?") {
-        ValidityResult::UnexpectedOutput(String::from(format!(
-            "{} \ngenerated singular query: {}",
-            res[0].as_str(),
-            query
-        )))
-    } else {
-        ValidityResult::Invalid(
-            None,
-            air::messages::error(
-                format!(
-                    "postcondition not satisfied: Ensures polynomial failed to be reduced to zero, reduced polynomial is {}\n generated singular query: {} ",
-                    res[0].as_str(),
-                    query
-                ),
-                &ens.1.spans[0],
-            ),
-        )
+
+    for stmt in &*stmts {
+        if let air::ast::StmtX::Assert(err, expr) = &**stmt {
+            assert_eq!(err.labels.len(), 1);
+            let span = &err.spans[0];
+            let label = &err.labels[0].note;
+            if label.contains("require") {
```

Comment:
We should not be using the error message label string to direct encoding choices, we should distinguish between them in the vir ast.

I think here https://github.com/verus-lang/verus/blob/main/source/vir/src/sst_to_air.rs#L2554-L2559 we are constructing the assumptions for the z3 query, and then we throw them away if it's a singular query. We may want to do something similar to what the z3 query does where the assumptions are passed in to the `QueryX` as locals.

Reply:
I updated how air asserts are added, so that the requires and ensures are separated into two blocks:
https://github.com/verus-lang/verus/blob/d86cf9f6ca18185ea12cf74b2acefef98d20866e/source/vir/src/sst_to_air.rs#L2473
although still implicit about the first block being the requires and second block being the ensures